Remove metaprogramming from Multilogger

miniengineer opened the pull request:
### Changes in this PR

1. replace `method_missing` metaprogramming with logging method definitions in `MultiLogger`
2. change order of expressions of message assignments in `FastLogger`
3. make `timestamp` the first key in `JSONLogger` message

### Background

#### 1
There are some cases when services check for logging method definitions before invoking them, e.g.

```
if logger.respond_to?(:info)
	# do some cool stuff
else
	# ooops
end
```

It breaks in case of `MultiLogger` since it doesn't implement logging methods, it uses metaprogramming to delegate them to logger instances.


#### 2

Some services use `progname` whereas Lorekeeper drops it, so the output response it not logged properly.
Instead we see

```
request
request
response
response
response
```


#### 3

Some messages may include time related keys, e.g. `last_updated_at`, which tracing services would use as a timestamp, instead of `timestamp` key that is present on the message object which leads to a discrepancy.
Changing the order seems to solve this.

### How to test

I have added specs that should be enough to test the changed behavior.
There is also a way to test all the changes by passing Lorekeeper as Faraday logger and making HTTP request.

You will need to ⬇️ 

1. `mkdir lorekeeper_playground`
2. `touch Gemfile` and add the following

```
/lorekeeper_playground/Gemfile

# frozen_string_literal: true

source "https://rubygems.org"

gem 'lorekeeper'
gem 'faraday'
```

3. `bundle`
4.  `touch logs.txt` (for writing JSON logs)
5.  `touch lorekeeper_playground.rb` and add the following

```
/lorekeeper_playground/lorekeeper_playground.rb

require 'bundler/setup'
require 'lorekeeper'
require 'faraday'

console_logger = Lorekeeper::SimpleLogger.new($stdout)
json_logger = Lorekeeper::JSONLogger.new('./logs.txt')

multi_logger = Lorekeeper::MultiLogger.new
multi_logger.add_logger(console_logger)
multi_logger.add_logger(json_logger)

conn = Faraday.new("http://httpbingo.org/") do |builder|
    builder.request :url_encoded
    builder.response :logger, multi_logger, { bodies: true, log_level: :debug }
    builder.adapter  :net_http
end

conn.get("/html")
```

6. `ruby lorekeeper_playground.rb`

You must see an error and no output, like so

```
/Users/yourusername/.rbenv/versions/2.7.4/lib/ruby/gems/2.7.0/gems/faraday-2.3.0/lib/faraday/logging/formatter.rb:102: warning:
Faraday::Logging::Formatter#debug at /Users/yourusername/.rbenv/versions/2.7.4/lib/ruby/2.7.0/forwardable.rb:154 forwarding to private method Lorekeeper::MultiLogger#debug
response
```

8. now edit the `Gemfile` so that `lorekepeer` source points to this branch

```
gem 'lorekeeper', git: "https://github.com/JordiPolo/lorekeeper", branch: "feature/remove-metaprogramming"
```

9. don't forget to `bundle` again 😉 
10. `ruby lorekeeper_playground.rb`

You must see logs output in the console + in the `logs.txt` file.
Make sure `timestamp` key comes first in `logs.txt` logs.

miniengineer commented on spec/lib/lorekeeper/multi_logger_spec.rb:
```diff
@@ -15,22 +15,22 @@
+      Lorekeeper::FastLogger::LOGGING_METHODS.each do |log_level|
+        logger.send(log_level, message) if logger.respond_to? (log_level)
 
-    it 'does not call the methods if they do not exist' do
```
I refactored the initial loggers added spec to be able to test the changes.
Maybe I misunderstand something, but I don’t think we were testing correctly before 🤔 

When we’re adding loggers to Multilogger, we don’t add the io (e.g. STDOUT) itself, instead we initialize either json or simple logger and pass an io as an argument. Only then we pass the instance of simple or json loggers to the Multilogger.

Tests are passing after the refactoring and I hope they make sense.

JordiPolo replied:
I think this was to ensure that if you do `logger.hello_world` somehow you did not end up with a log message. if we are removing method_missing I think this test is meaningless

miniengineer commented on lib/lorekeeper/multi_logger.rb:
```diff
@@ -26,7 +37,7 @@ def respond_to?(method)
     end
 end
 
-    def method_missing(method, *args, &block)
```
After I renamed the  method_missing method two specs started failing
[One of them](https://github.com/JordiPolo/lorekeeper/blob/d3bcff1ae3c6600cafc79bec34cd801d4ea1b3a9/spec/lib/lorekeeper/multi_logger_spec.rb#L24) checks whether `#write`  method can be called from Multilogger.

But `#write` belongs to a private class within `FastLogger` and the method is called within the `#add` method.

I believe the `#add` method is the main one to implement and it’s the entry point for any logger class, whereas `#write` and other methods are not meant to be called by from outside.

So, I kept the `method_missing` renamed (as suggested) and I removed [the spec that checks](https://github.com/JordiPolo/lorekeeper/pull/26/files#diff-3ec923edcde18eaf057a5e7a8de440fd02df05fac0b2e63022437bc8a83d684fL24) whether we can call `#write` method.

ykitamura-mdsol replied:
I don't think we should drop the `write` method support. Some loggers need it.

miniengineer commented:
@jfeltesse-mdsol 

thank you for the review 🙏 
added the suggested changes

I took the liberty of disabling the `no-single-line-method` in `.rubocop.yml` and added related TODO there
hope it's ok 🤔

miniengineer commented:
btw, while am I at it is it ok to add running `rubocop` as part of the `build` script?

We only run `rspec` tests :O 

https://github.com/JordiPolo/lorekeeper/blob/d3bcff1ae3c6600cafc79bec34cd801d4ea1b3a9/.github/workflows/build.yml#L26

JordiPolo commented on lib/lorekeeper/fast_logger.rb:
```diff
@@ -56,7 +56,7 @@ def initialize(file)
     def add(severity, message_param = nil, progname = nil, &block)
       return true if severity < @level
-      message = message_param || (block && block.call) || progname
+      message = (block && block.call) || message_param || progname
```
why the switch here?

ykitamura-mdsol replied:
it seems some handsome guy suggested it

<img width="943" alt="Screen Shot 2022-07-27 at 5 29 13 PM" src="https://user-images.githubusercontent.com/16286071/181376155-c5a02c71-c697-4714-9452-b3230e18c2d8.png">

JordiPolo replied:
Very handsome indeed

JordiPolo commented on lib/lorekeeper/multi_logger.rb:
```diff
@@ -16,17 +16,38 @@ def inspect
+
+    def fatal_with_data(*args, &block); call_loggers(:fatal, *args, &block); end
+
+    if RUBY_VERSION > '2.6'
```
@ykitamura-mdsol at this stage, can't we asume we are at least in 2.6 ?

ykitamura-mdsol replied:
yeah, I think it's safe to drop ruby 2.5 support

https://github.com/JordiPolo/lorekeeper/blob/d3bcff1ae3c6600cafc79bec34cd801d4ea1b3a9/lorekeeper.gemspec#L23

ykitamura-mdsol commented:
> btw, while am I at it is it ok to add running `rubocop` as part of the `build` script?

why not, go for it!

miniengineer commented:
> > btw, while am I at it is it ok to add running rubocop as part of the build script?

>why not, go for it!



Awesome, I have a draft here 🎉 
being a guest in his repo 🙈 , will need to sync with Jordi later for the rules

miniengineer commented:
@ykitamura-mdsol  @JordiPolo 
thank you for the reviews 🙏 

Just one question ❓ 
When I added back the support for `#write` method in Multilogger, having a test like this 

```
    it 'calls write method of loggers' do
      logger.add_logger(console_logger)

      logger.write(message)
      expect(io.received_message).to eq(message)
    end
```

and debugging with `byebug` revealed an error like this ⬇️ 

```
*** NoMethodError Exception: undefined method `write' for #<Lorekeeper::SimpleLogger:0x00007f7b48b95508>
```

Looked at the previous test and there it passed an io instance instead of logger.
Refactored and it works 👍 

Just to make sure I make a correct assumption
Since calling `#write` to Multilogger instance is a common for some loggers, as mentioned by Yohei,
it's ok to pass just an IO to Multilogger, it doesn't have to be an instance of a logger? 🤔 

thank you!!

jfeltesse-mdsol commented:
you need to

[ ] update gemspec to make minimum ruby version 2.7.0
[ ] bump the version number
[ ] update the changelog

miniengineer commented:
@JordiPolo 

just to be safe, is it OK to drop the support to any ruby version that is less than `2.7`? 🙏 
thank you

ykitamura-mdsol commented:
> Is it ok to pass just an IO to Multilogger, it doesn't have to be an instance of a logger? 🤔

It's okay and supporting `write` in SimpleLogger is also something we want to do #7

ykitamura-mdsol reviewed (approved):
Good job! Looks good to me 👍
